packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,307 @@
+/**
+ * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module source-editing/sourceediting
+ */
+
+import { Plugin, PendingActions } from 'ckeditor5/src/core';
+import { ButtonView } from 'ckeditor5/src/ui';
+import { createElement, ElementReplacer } from 'ckeditor5/src/utils';
+
+import '../theme/sourceediting.css';
+
+// TODO: create icon
+// import sourceEditingIcon from '../theme/icons/sourceediting.svg';
+
+const COMMAND_FORCE_DISABLE_ID = 'SourceEditingMode';
+
+/**
+ * The source editing feature.
+ *
+ * It provides the possibility to view and edit the source of the document.
+ *
+ * For a detailed overview, check the {@glink features/source-editing source editing feature documentation} and the
+ * {@glink api/source-editing package page}.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class SourceEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'SourceEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Flag indicating whether the document source mode is active.
+		 *
+		 * @observable
+		 * @member {Boolean}
+		 */
+		this.set( 'isSourceEditingMode', false );
+
+		/**
+		 * The element replacer instance used to replace the editing roots with the wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {module:utils/elementreplacer~ElementReplacer}
+		 */
+		this._elementReplacer = new ElementReplacer();
+
+		/**
+		 * Maps all root names to wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {Map.<String,HTMLElement>}
+		 */
+		this._replacedRoots = new Map();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const t = editor.t;
+
+		editor.ui.componentFactory.add( 'sourceEditing', locale => {
+			const buttonView = new ButtonView( locale );
+
+			buttonView.set( {
+				label: t( 'Edit source' ),
+				// TODO: use icon
+				// icon: sourceEditingIcon,
+				tooltip: true,
+				withText: true
+			} );
+
+			buttonView.bind( 'isOn' ).to( this, 'isSourceEditingMode' );
+
+			// Disable button if:
+			// - plugin is disabled, or
+			// - editor is in a read-only mode, or
+			// - there is a pending action.
+			//
+			// Pending action may change the model, so viewing and/or editing the document source should be prevented until the model is
+			// finally set.
+			buttonView.bind( 'isEnabled' ).to(
+				this, 'isEnabled',
+				editor, 'isReadOnly',
+				editor.plugins.get( PendingActions ), 'hasAny',
+				( isEnabled, isEditorReadOnly, hasAnyPendingActions ) => isEnabled && !isEditorReadOnly && !hasAnyPendingActions
+			);
```

Comment:
```suggestion
			// The button should be disabled if one of the following condition is met.
			buttonView.bind( 'isEnabled' ).to(
				this, 'isEnabled',
				editor, 'isReadOnly',
				editor.plugins.get( PendingActions ), 'hasAny',
				( isEnabled, isEditorReadOnly, hasAnyPendingActions ) => {
					// (1) The plugin is disabled itself.
					if ( !isEnabled ) {
						return false;
					}

					// (2) The editor is in read-only mode.
					if ( isEditorReadOnly ) {
						return false;
					}

					// (3) Any pending action is scheduled. It may change the model, so modifying the document source 
					// should be prevented until the model is finally set.
					if ( hasAnyPendingActions ) {
						return false;
					}

					return true;
				}
			);
```

WDYT?

Reply:
Good idea, thanks.